dbms/src/Storages/KVStore/Decode/DecodingStorageSchemaSnapshot.cpp

```diff
@@ -17,6 +17,94 @@
 
 namespace DB
 {
+DecodingStorageSchemaSnapshot::DecodingStorageSchemaSnapshot(
+    DM::ColumnDefinesPtr column_defines_,
+    const TiDB::TableInfo & table_info_,
+    const DM::ColumnDefine & original_handle_,
+    Int64 decoding_schema_epoch_,
+    bool has_version_column)
+    : column_defines{std::move(column_defines_)}
+    , pk_is_handle{table_info_.pk_is_handle}
+    , is_common_handle{table_info_.is_common_handle}
+    , decoding_schema_epoch{decoding_schema_epoch_}
+{
+    std::unordered_map<ColumnID, size_t> column_lut;
+    // col id -> tidb pos, has no internal cols.
+    for (size_t i = 0; i < table_info_.columns.size(); i++)
+    {
+        const auto & ci = table_info_.columns[i];
+        column_lut.emplace(ci.id, i);
+    }
+    // column_defines has internal cols.
+    size_t index_in_block = 0;
+    for (size_t i = 0; i < column_defines->size(); i++)
+    {
+        auto & cd = (*column_defines)[i];
+        if (cd.id != VersionColumnID || has_version_column)
+        {
+            col_id_to_block_pos.insert({cd.id, index_in_block++});
+        }
+        col_id_to_def_pos.insert({cd.id, i});
```

Comment:
ditto, reserve first

Reply:
Sorry, should have comment here https://stackoverflow.com/questions/42373232/why-does-stdunordered-map-have-a-reserve-method